Hash my url

alishaz-polymath opened the pull request:
## What does this PR do?

**Adds**:
- a new path `/d/[link]` for temporary links (can be used for disposable booking links when that feature is implemented)
- a new booking page `/d/book` for booking using temporary links (can be used for disposable booking links when that feature is implemented)
- adds a new model for hashed links/urls
```
model HashedLink {
  id          Int       @id @default(autoincrement())
  link        String    @unique()
  eventType   EventType @relation(fields: [eventTypeId], references: [id], onDelete: Cascade)
  eventTypeId Int       @unique
}
```

**Feature**:
- Users can enable a setting in the event-type which creates a hash URL they can share with the clients they don't want to reveal their usernames/slugs
- Once the booking is complete using this link, the link automatically expires and a new one is generated in its place
- The regular event-type link works normally, this hashed URL is usable as a separate feature for the same event-type

**Demo**:
[Loom Video](https://www.loom.com/share/0a91d90784344a0184009097d3a485a3)

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

Tried pushing changes to `staging` (app.cal.dev) but deployment failed due to unrelated error.

Can test on staging with the specific preview deployment with just a little bit of hassle:
[Loom Here](https://www.loom.com/share/198da89aefca498895b5726f68173a94)

Can also be tested locally by pulling the branch. Follow the steps in loom [here](https://www.loom.com/share/0a91d90784344a0184009097d3a485a3)

Playwright tests have been added to confirm that 
- [x] Hashed URL is being created successfully
- [x] Hashed URL is being used to book successfully
- [x] Upon booking, a new hashed URL is being generated which replaces the previous one

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- This PR needs changes to the documentation

alishaz-polymath committed:
disposable link model added

alishaz-polymath committed:
disposable model updated

alishaz-polymath committed:
added disposable slug availability page

alishaz-polymath committed:
added disposable book page

alishaz-polymath committed:
added disposable slug hook

alishaz-polymath committed:
added disposable link booking flow

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
updated schema

alishaz-polymath committed:
checktype fix

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
added checkfix and schema generated

alishaz-polymath committed:
create link API added

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
added one time link view on event type list

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
adjusted schema

alishaz-polymath committed:
fixed disposable visual indicator

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
expired check and visual indicator added

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
updated slug for disposable event type

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
revised schema

alishaz-polymath committed:
WIP

alishaz-polymath committed:
revert desc

alishaz-polymath committed:
revert --WIP

alishaz-polymath committed:
rework based on change of plans

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
further adjustments

alishaz-polymath committed:
added eventtype option for hashed link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
added refresh and delete on update

alishaz-polymath committed:
fixed update call conditions

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
cleanup

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
code improvement

alishaz-polymath commented on apps/web/pages/[user]/book.tsx:
```diff
@@ -196,6 +196,8 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
       booking,
       trpcState: ssr.dehydrate(),
       isDynamicGroupBooking,
+      hasHashedBookingLink: false,
+      hashedLink: undefined,
```
To confirm the possible associated hashed link shouldn't be updated after the event is booked

alishaz-polymath commented on apps/web/pages/api/book/event.ts:
```diff
@@ -772,6 +773,20 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
       },
     },
   });
+  // refresh hashed link if used
+  const urlSeed = `${users[0].username}:${dayjs(req.body.start).utc().format()}`;
+  const hashedUid = translator.fromUUID(uuidv5(urlSeed, uuidv5.URL));
+
+  if (hasHashedBookingLink) {
+    await prisma.hashedLink.update({
+      where: {
+        link: reqBody.hashedLink as string,
+      },
+      data: {
+        link: hashedUid,
+      },
+    });
+  }
```
Hashed link is updated if used for booking

alishaz-polymath commented on apps/web/pages/event-types/[type].tsx:
```diff
@@ -442,6 +445,10 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
     team ? `team/${team.slug}` : eventType.users[0].username
   }/${eventType.slug}`;
 
+  const placeholderHashedLink = `${process.env.NEXT_PUBLIC_WEBSITE_URL}/d/${
+    eventType.hashedLink ? eventType.hashedLink.link : "xxxxxxxxxxxxxxxxx"
+  }/${eventType.slug}`;
```
The placeholder hashed link to show in the input field and for copying purposes, if the hashed link is already generated, else a fallback placeholder which won't be copied until saved/updated

alishaz-polymath commented on apps/web/pages/event-types/[type].tsx:
```diff
@@ -1330,6 +1338,65 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
+                                }}
+                              />
+                              {hashedLinkVisible && (
+                                <div className="block items-center sm:flex">
+                                  <div className="min-w-48 mb-4 sm:mb-0"></div>
+                                  <div className="w-full">
+                                    <div className="relative mt-1 flex w-full">
+                                      <input
+                                        disabled
+                                        type="text"
+                                        className="  grow select-none border-gray-300 bg-gray-50 text-sm text-gray-500 ltr:rounded-l-sm rtl:rounded-r-sm"
+                                        placeholder={t("meeting_with_user")}
+                                        defaultValue={placeholderHashedLink}
+                                      />
+                                      <Tooltip
+                                        content={
+                                          eventType.hashedLink
+                                            ? t("copy_to_clipboard")
+                                            : t("enabled_after_update")
+                                        }>
+                                        <Button
+                                          color="minimal"
+                                          onClick={() => {
+                                            if (eventType.hashedLink) {
+                                              navigator.clipboard.writeText(placeholderHashedLink);
+                                              showToast("Link copied!", "success");
+                                            }
+                                          }}
+                                          type="button"
+                                          className="text-md flex items-center border border-gray-300 px-2 py-1 text-sm font-medium text-gray-700 ltr:rounded-r-sm ltr:border-l-0 rtl:rounded-l-sm rtl:border-r-0">
+                                          <DocumentDuplicateIcon className="w-6 p-1 text-neutral-500" />
+                                        </Button>
+                                      </Tooltip>
+                                    </div>
+                                    <span className="text-xs text-gray-500">
+                                      The URL will regenerate after each use
+                                    </span>
+                                  </div>
+                                </div>
+                              )}
```
Show hashed link placeholder if the "Generate hash link" field is checked

alishaz-polymath commented on apps/web/server/routers/viewer/eventTypes.tsx:
```diff
@@ -234,6 +246,48 @@ export const eventTypesRouter = createProtectedRouter()
+      if (hashedLink) {
+        // check if hashed connection existed. If it did, do nothing. If it didn't, add a new connection
+        if (!connectedLink) {
+          const translator = short();
+          const seed = `${input.eventName}:${input.id}:${new Date().getTime()}`;
+          const uid = translator.fromUUID(uuidv5(seed, uuidv5.URL));
+          // create a hashed link
+          await ctx.prisma.hashedLink.upsert({
+            where: {
+              eventTypeId: input.id,
+            },
+            update: {
+              link: uid,
+            },
+            create: {
+              link: uid,
+              eventType: {
+                connect: { id: input.id },
+              },
+            },
+          });
+        }
```
Create a new link if there was no prior connected link. This is done if the "Generate hash link" was checked in event type update.

alishaz-polymath commented on apps/web/server/routers/viewer/eventTypes.tsx:
```diff
@@ -234,6 +246,48 @@ export const eventTypesRouter = createProtectedRouter()
+        }
+      } else {
+        // check if hashed connection exists. If it does, disconnect
+        if (connectedLink) {
+          await ctx.prisma.hashedLink.delete({
+            where: {
+              eventTypeId: input.id,
+            },
+          });
+        }
```
Disconnect any connected links if a connection exists and the "Generate Hashed link" field was checked in the event type update.

alishaz-polymath reviewed (commented):
Self review added

alishaz-polymath committed:
clean up

hariombalhara commented:
I noticed that one disposable link is shown at a time to the user. If a user shares it with 2 persons, only one of them can successfully book the meeting.
Shouldn't we allow to generate multiple links simulatenously to solve this problem ?

alishaz-polymath commented:
> I noticed that one disposable link is shown at a time to the user. If a user shares it with 2 persons, only one of them can successfully book the meeting. Shouldn't we allow to generate multiple links simulatenously to solve this problem ?

That is by design. The use-case for multiple bookings with the same link will be taken care of in the "disposable booking links" feature, which solves for a different use case. The use-case in this feature is specifically meant for a one-off single use booking where the cal username is not shared with the clients

hariombalhara committed:
Potential fix for 404

alishaz-polymath committed:
backward compat for booking page

alishaz-polymath committed:
fixes regular booking for user and team

alishaz-polymath committed:
typefix

zomars commented:
Minor nitpick: We keep calling it "disposable link" but we named it "hashed link" in codee. Shouldn't we just name it `DisposableLink` in the code?

alishaz-polymath commented:
> Minor nitpick: We keep calling it "disposable link" but we named it "hashed link" in code. Shouldn't we just name it `DisposableLink` in the code?

After a lot of discussions between @Jaibles and I, and @Jaibles , @PeerRich and @baileypumfleet  it was concluded that we'd reserve disposable link for a follow up feature which solves for the issue [Here](https://github.com/calcom/cal.com/issues/1474)

zomars commented on apps/web/components/booking/pages/BookingPage.tsx:
```diff
@@ -12,6 +12,7 @@ import { useSession } from "next-auth/react";
 import dynamic from "next/dynamic";
 import Head from "next/head";
 import { useRouter } from "next/router";
+import { HashLinkPageProps } from "pages/d/[link]/book";
```
Do we mean to use an absolute import here?

alishaz-polymath replied:
Oh, right. Updating this to be similar to `BookPageProps` & `TeamBookingPageProps`

zomars reviewed (changes requested):
Code looks good! Since this is a new feature I would require two things:
1. Get this working on staging
2. Add E2E to make sure this feature works

alishaz-polymath committed:
updated path for Booking import

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
checkfix

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

zomars commented:
> 1. This works well on staging but on the specific deployment ("https://cal-q7o6xb2oa-cal.vercel.app", for example). I think this doesn't work on `dev.cal.com` yet because of migrations?

I think you mean `app.cal.dev`. Both Vercel preview deployments and this share the same DB. So if one has a migration means the other one as well.

> 2. For E2E test, the above issue might result in failed test, right? 

Not really since tests don't run on Vercel and are ran on a fresh DB.

> We would probably need the migration to be set in staging before the E2E test can actually run fine because it would expect to see this feature functional in staging, right?

Nope as explained above. E2E tests are run on a fresh environment.

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath commented:
> > 1. This works well on staging but on the specific deployment ("https://cal-q7o6xb2oa-cal.vercel.app", for example). I think this doesn't work on `dev.cal.com` yet because of migrations?
> 
> I think you mean `app.cal.dev`. Both Vercel preview deployments and this share the same DB. So if one has a migration means the other one as well.
> 

@zomars

So, That's weird because I see the changes here (notice the `Generate hashed URL` checkbox):
![image](https://user-images.githubusercontent.com/52925846/165255281-39c93764-64cd-4475-a86e-9b2ffd1abcd7.png)

But not here:
![image](https://user-images.githubusercontent.com/52925846/165255064-8f5e1235-8e48-4eed-8656-44ff44d99883.png)

Is there a reason why this might be? Or do I have to manually push a particular branch into dev deployment?

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
e2e wip

alishaz-polymath committed:
Merge branch 'feature/disposable-booking-link' of https://github.com/calcom/cal.com into feature/disposable-booking-link

alishaz-polymath committed:
link err fix

zomars commented:
You need to push to the `staging` branch.

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
workaround for banner issue in event type update-test

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
added regenerate hash check

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
Merge branch 'main' into feature/disposable-booking-link

alishaz-polymath committed:
fixed test according to new testID

alishaz-polymath commented:
@zomars I have added the 

> Code looks good! Since this is a new feature I would require two things:
> 
>     1. Get this working on staging
> 
>     2. Add E2E to make sure this feature works

1. I pushed changes to staging but deployment fails (due to unrelated issue) and hence it can't be checked on `app.cal.dev` however, it works like a charm on preview deployments in staging. [Here's a loom](https://www.loom.com/share/198da89aefca498895b5726f68173a94) of how you can test it on preview deployments as well

2. Added E2E tests and they work

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link

zomars committed:
Merge branch 'main' into feature/disposable-booking-link

zomars reviewed (approved):
Functionally it works like a charm. I'm still not 100% in the UX side of things but I'll let @Jaibles be the judge of that and we can follow up later. Thanks @alishaz-polymath 🙏🏽

kodiakhq[bot] committed:
Merge branch 'main' into feature/disposable-booking-link
